File: app/src/main/java/com/goblinvault/ExpenseManager.java
package com.goblinvault;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ExpenseManager {

    private List<Expense> expenses;

    public ExpenseManager() {
        this.expenses = new ArrayList<>();
    }

    public void addExpense(Expense expense) {
        expenses.add(expense);
    }

    public List<Expense> getExpensesForYear(int year) {
        List<Expense> filteredExpenses = new ArrayList<>();
        for (Expense expense : expenses) {
            if (expense.getYearTag() == year) {
                filteredExpenses.add(expense);
            }
        }
        return filteredExpenses;
    }

    public List<Expense> getPieChartForYear(int year) {
        return calculatePieChart(getExpensesForYear(year));
    }

    public List<Expense> getExpensesForMonth(int year, int month) {
        List<Expense> filteredExpenses = new ArrayList<>();
        for (Expense expense : expenses) {
            if (expense.getYearTag() == year && expense.getMonthTag() == month) {
                filteredExpenses.add(expense);
            }
        }
        return filteredExpenses;
    }

    public List<Expense> getPieChartForMonth(int year, int month) {
        return calculatePieChart(getExpensesForMonth(year, month));
    }

    public List<Expense> getExpensesForDay(int year, int month, int day) {
        List<Expense> filteredExpenses = new ArrayList<>();
        for (Expense expense : expenses) {
            if (expense.getYearTag() == year && expense.getMonthTag() == month && expense.getDayTag() == day) {
                filteredExpenses.add(expense);
            }
        }
        return filteredExpenses;
    }

    public List<Expense> getPieChartForDay(int year, int month, int day) {
        return calculatePieChart(getExpensesForDay(year, month, day));
    }

    private List<Expense> calculatePieChart(List<Expense> filteredExpenses) {
        Map<ExpenseCategory, Float> categoryTotalMap = new HashMap<>();

        for (Expense expense : filteredExpenses) {
            ExpenseCategory category = expense.getCategory();
            float totalAmount = categoryTotalMap.getOrDefault(category, 0f);
            categoryTotalMap.put(category, totalAmount + expense.getAmount());
        }

        List<Expense> pieChartExpenses = new ArrayList<>();
        for (Map.Entry<ExpenseCategory, Float> entry : categoryTotalMap.entrySet()) {
            ExpenseCategory category = entry.getKey();
            float totalAmount = entry.getValue();
            pieChartExpenses.add(new Expense(totalAmount, category, System.currentTimeMillis(), 0, 0, 0));
        }

        pieChartExpenses.sort((expense1, expense2) -> Float.compare(expense2.getAmount(), expense1.getAmount()));
        return pieChartExpenses;
    }
}

File: app/src/main/java/com/seedling/demo/hostdeveloperdemo/ExpenseManager.kt
package com.seedling.demo.hostdeveloperdemo

class ExpenseManager {

    private val expenses: MutableList<Expense> = mutableListOf()

    fun addExpense(expense: Expense) {
        expenses.add(expense)
    }

    fun getExpensesForYear(year: Int): List<Expense> {
        return expenses.filter { it.yearTag == year }
    }

    fun getPieChartForYear(year: Int): List<Expense> {
        return calculatePieChart(expenses.filter { it.yearTag == year })
    }

    fun getExpensesForMonth(year: Int, month: Int): List<Expense> {
        return expenses.filter { it.yearTag == year && it.monthTag == month }
    }

    fun getPieChartForMonth(year: Int, month: Int): List<Expense> {
        return calculatePieChart(expenses.filter { it.yearTag == year && it.monthTag == month })
    }

    fun getExpensesForDay(year: Int, month: Int, day: Int): List<Expense> {
        return expenses.filter { it.yearTag == year && it.monthTag == month && it.dayTag == day }
    }

    fun getPieChartForDay(year: Int, month: Int, day: Int): List<Expense> {
        return calculatePieChart(expenses.filter { it.yearTag == year && it.monthTag == month && it.dayTag == day })
    }

    private fun calculatePieChart(filteredExpenses: List<Expense>): List<Expense> {
        val categoryTotalMap = mutableMapOf<ExpenseCategory, Float>()

        filteredExpenses.forEach { expense ->
            val totalAmount = categoryTotalMap.getOrDefault(expense.category, 0f)
            categoryTotalMap[expense.category] = totalAmount + expense.amount
        }

        // Convert the map back to a list of Expense objects
        return categoryTotalMap.map { (category, totalAmount) ->
            Expense(totalAmount, category, System.currentTimeMillis(), 0, 0, 0)
        }.sortedByDescending { it.amount }
    }
}